buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle.reaper;
+
+import java.io.ByteArrayOutputStream;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A standalone process that will reap external services after a build dies.
+ *
+ * <h2>Input</h2>
+ * Since how to reap a given service is platform and service dependent, this tool
+ * operates on system commands to execute. It takes a single argument, a directory
+ * that will contain files with reaping commands. Each line in each file will be
+ * executed with {@link Runtime#getRuntime()#exec}.
+ *
+ * The main method will wait indefinitely on the parent process (Gradle) by
+ * reading from stdin. When Gradle shuts down, whether normally or abruptly, the
+ * pipe will be broken and read will return.
+ *
+ * The reaper will then iterate over the files in the configured directory,
+ * and execute the given commands. If any commands fail, a failure message is
+ * written to stderr. Otherwise, the input file will be deleted. If no inputs
+ * produced errors, the entire input directory is deleted upon completion of reaping.
+ */
+public class Reaper implements Closeable {
+
+    Path inputDir;
+    boolean failed;
+
+    private Reaper(Path inputDir) {
+        this.inputDir = inputDir;
+        this.failed = false;
+    }
+
+    public static void main(String[] args) throws Exception {
+        if (args.length != 1) {
+            System.err.println("Expected one argument.\nUsage: java -cp reaper.jar Reaper <DIR_OF_REAPING_COMMANDS>");
+            System.exit(1);
+        }
+        Path inputDir = Paths.get(args[0]);
+
+        try (Reaper reaper = new Reaper(inputDir)){
+            System.in.read();
+            reaper.reap();
+        }
+    }
+
+    private void reap() {
+        try {
```

Comment:
I think we might want to make this exception handling a bit more granular. For example, if we blow up trying to kill one process the check in `close()` will mean we don't clean up any of these command files. Later if we try to run this again calling `kill` against an already killed pid will fail, which will mean `failed == false` and this cycle continues. 

To completely get around this we'll probably have to delete the command files as we successfully execute them such that if we blow up mid-way we can pick up where we left off next time. We still might want to do per-command exception handling though such that we can clean up as much as we can. We don't want to keep any more stuff dangling than we have to for whatever reason.

Reply:
We read the files up front, before any processing, specifically so that we don't rely on the state of the directory changing. There isn't any possibility for a failure cycle.

While it is possible we could blow up when trying to log a failure of one of the commands (getting an ioexception when writing to the log), we are pretty hosed at that point, so I'm not sure special handling is really worthwhile, since some investigation would need to be done anyways. There is no "next time" with this, since the directory is per gradle daemon. The only way the reaper should actually do work is if the daemon died (since we should be unregistering everything in the normal case).

Reply:
> We read the files up front, before any processing, specifically so that we don't rely on the state of the directory changing. 

We only read the _list_ of files. What I'm saying is, what if I have 5 processes to kill and something happens on say, the third one? Right now we would immediately stop, not even attempting to kill the final 2, and mark reaping as having failed. On `close()` we would then not delete _any_ of the pid files meaning the next time we go to reap we would immediately fail again on the first file because the process _had_ been reaped, but we haven't marked it as such.

Reply:
> The only way the reaper should actually do work is if the daemon died (since we should be unregistering everything in the normal case).

Not necessarily, we could fail to stop these processes gracefully for any number of reasons that doesn't necessarily involve the Gradle daemon dying.

Reply:
> we could fail to stop these processes gracefully for any number of reasons that doesn't necessarily involve the Gradle daemon dying.

This shoudln't be possible. Testclusters falls back to using SIGKILL if SIGTERM doesn't work, and if SIGKILL somehow fails, it fails the build.

>  meaning the next time we go to reap

There is no next time. The reaper is per build, and it only does work at the end of the build, when gradle died.

Reply:
The directory with the files is based on the daemon pid though. Isn't it possible the next build with the same daemon will run a reaper pointed at the same directory?

Reply:
We delete the directory completely, and in this case the *only* thing that could fail is deleting our log file just before deleting the dir (since those are the only two files when we exited normally). But if that happens we have bigger problems (if we can't delete, then we likely have permission problems, which will persistent regardless). So the next build with the same daemon will create a fresh copy of the dir.

Reply:
We only delete the directory if we don't encounter any failures. As I described above, if any of our reaping commands fail, we don't clean up any of this stuff. There's any number of reasons why this might happen (processes crash, are killed manually outside Gradle, etc). In such a case so long as that daemon lives we'll just keep trying and trying, and failing and failing. At the very least we should delete these files as we go, such that when we do successfully reap a process we immediately clean up the pid file.

Reply:
> As I described above, if any of our reaping commands fail, we don't clean up any of this stuff. 

If there are any reaping commands to run, then gradle died. We should be unregistering all commands from the reaper in any build that completes, whether a success or failure.

I have made the delete happen on successful execution of the command. If we end up having problems for some reason, we can randomize this directory to make it unique per build instance instead of daemon, but I think that can be a followup, and only if we run into issues.

Reply:
> If there are any reaping commands to run, then gradle died. We should be unregistering all commands from the reaper in any build that completes, whether a success or failure.

Ah, understood.

I wonder if we are going to need anything to periodically clean these directories up? It's very possible that we later launch a daemon with the same pid, which might in turn have stale stuff in there.

Reply:
Since the directory is deleted in the normal case, I think leaving it to the user to cleanup if old directories exist is fine. Adding random element to the dir name as I described above would stop a future daemon from reusing a potentially failed daemon's dir.

Reply:
Yeah, my only concern is that if we ran into a situation where we were using an old daemon dir it would be hard for a user to figure that out. I'd say the scenario of reusing an old pid that _also_ failed to reap processes is fairly low.

Going forward the preferred method would be some kind of protocol that we communicate with the reaper with over stdin. That way the reaper process itself holds all this state and we can ditch all this filesystem stuff. We can tackle that another day though.